tests/crowdsourcing/tasks/test_chat_demo.py

```diff
@@ -0,0 +1,498 @@
+
+            # Have agents fill out the form
+            self.server.send_agent_act(
+                agent_id=agent_0_id,
+                act_content={
+                    'text': FORM_PROMPTS['agent_0'],
+                    'task_data': {'form_responses': FORM_RESPONSES['agent_0']},
+                    'id': AGENT_0_DISPLAY_ID,
+                    'episode_done': False,
+                },
+            )
+            self.server.send_agent_act(
+                agent_id=agent_1_id,
+                act_content={
+                    'text': FORM_PROMPTS['agent_1'],
+                    'task_data': {'form_responses': FORM_RESPONSES['agent_1']},
+                    'id': AGENT_1_DISPLAY_ID,
+                    'episode_done': False,
+                },
+            )
+
+            # Submit the HIT
+            self.server.send_agent_act(
+                agent_id=agent_0_id,
+                act_content={
+                    'task_data': {'final_data': {}},
+                    'MEPHISTO_is_submit': True,
+                },
+            )
+            self.server.send_agent_act(
+                agent_id=agent_1_id,
+                act_content={
+                    'task_data': {'final_data': {}},
+                    'MEPHISTO_is_submit': True,
+                },
+            )
+
+            # # Check that the inputs and outputs are as expected
+
+            state_0, state_1 = [
+                agent.state.get_data() for agent in self.db.find_agents()
+            ]
+            actual_and_desired_states = [
+                (state_0, DESIRED_STATE_AGENT_0),
+                (state_1, DESIRED_STATE_AGENT_1),
+            ]
+            for actual_state, desired_state in actual_and_desired_states:
+                assert actual_state['inputs'] == desired_state['inputs']
+                assert len(actual_state['outputs']['messages']) == len(
+                    desired_state['outputs']['messages']
+                )
+                for actual_message, desired_message in zip(
+                    actual_state['outputs']['messages'],
+                    desired_state['outputs']['messages'],
+                ):
+                    for key, desired_value in desired_message.items():
+                        if key == 'timestamp':
+                            pass  # The timestamp will obviously be different
+                        elif key == 'data':
+                            for key_inner, desired_value_inner in desired_message[
+                                key
+                            ].items():
+                                if key_inner == 'message_id':
+                                    pass  # The message ID will be different
+                                else:
+                                    self.assertEqual(
+                                        actual_message[key][key_inner],
+                                        desired_value_inner,
+                                    )
```

Comment:
This will likely be the kind of thing to refactor out in the future, as this is likely to be the same comparator function you'll use for most tasks (perhaps specifying subkeys that should be ignored?), but it's valuable to have in a working example so far!

Reply:
Oh yeah, definitely - most of this should be factored out as we build out more similar tests :)